### PR TITLE
Add rewrite packed structs pass

### PR DESCRIPTION
**Pull request by omarahmed1111**
# Motivation
packed structs would get the alignment of its biggest size type, so when we pass them as input buffers. they need to align with its total size as we will need to store them in runtime arrays which we will need to assign an array stride for them. eg: <{i32, i8}> struct will have an alignment of 4 and a size of 5, when we store it inside a runtime array, we need to assign an array stride of 5 to the runtime array. This will fail in validation as the array stride(5) need to align with struct alignment(4).

# PR changes
* Add `RewritePackeStructsPass` which will reduce alignment of packed structs by converting the input buffer struct type to a struct of array of i8 with equal size. This pass handles opaque and non-opaque pointers:
  * For opaque pointers: We store a zero values of the equivalent struct type inside the input buffer ptr (We do that to make opaque pointer type deduced as of the type that we need). We also allocate a local ptr of the same original struct type of the input buffer, and replace all uses inside the kernel with our allocated ptr. At the end of the function, we extract each 8 bits from our allocated ptr and assign this 8 bits to the input buffer i8 array.
  * For non-opaque pointers: We change the signature of the kernel function and replace the input buffer type to our new struct type. We bitcast the input buffer from the new struct type to the original struct type. And, we replace all uses of the input buffer inside the function to the bitcasted value.
* Add `-rewrite-packed-struct` option to trigger the pass.
* Add some tests for the pass.

Note: We need to have `-int8` option enabled to process the pass.

This fixes the CTS test: spirv_new/decorate_cpacked

**This contribution is being made by Codeplay on behalf of Samsung.**

### CODE REVIEW

File: lib/Option.cpp [hunk 1]
```diff
@@ -110,6 +110,12 @@ llvm::cl::opt<bool> hack_inserts(
         "into struct types by using complete composite construction and "
         "extractions"));
 
+llvm::cl::opt<bool> rewrite_packed_structs(
+    "rewrite-packed-structs", llvm::cl::init(false),
+    llvm::cl::desc(
+        "Rewrite packed structs passed as buffers to a new packed structs with "
+        "an array of i8 of equal size to reduce struct alignment"));
+
```
[the remark this answers is not in the document]
Comment (rjodinchr) [reply lib/Option.cpp:118]: I prefer the way it is now with the `llvm_unreachable`. It forces users to be aware of it.

**Comment by rjodinchr** [comment]
Should we also consider doing the same thing for structure on which we tried to store an element smaller than its element size?
At the moment, it emits a `llvm_unreachable` (https://github.com/google/clspv/blob/main/lib/ReplacePointerBitcastPass.cpp#L429) when we try to do that. I don't think I have seen it in the CTS yet.

**Review by alan-baker — COMMENTED** [review]
Have you tested the interactions with `-constant-args-ubo`? Would it simply be a case of requiring scalar block layout to work?

**Comment by omarahmed1111** [comment]
> Have you tested the interactions with -constant-args-ubo?  

I tried testing that but as of my understanding that `-constant-args-ubo` works with constant buffers and add some padding to structs to align them properly and if constant args are passed in a wrong way would give that frontend error: `clspv restriction: UBO element size must be a multiple of that element's alignment.` which this restriction would make `rewrite-packed-structs` not work on the struct as the struct will be already properly aligned with its size. so, I guess there is no conflict between the 2 options.

> Would it simply be a case of requiring scalar block layout to work?

Scalar block layout should still give the struct the alignment of the biggest type inside of it. which would still conflict with the arraystride of the runtime array that we assign as a wrapper to the buffer. So, as of my understanding it won't fix the problem of having a packed struct that is not aligned with its size.

**Comment by omarahmed1111** [comment]
> Should we also consider doing the same thing for structure on which we tried to store an element smaller than its element size?
At the moment, it emits a llvm_unreachable (https://github.com/google/clspv/blob/main/lib/ReplacePointerBitcastPass.cpp#L429) when we try to do that. I don't think I have seen it in the CTS yet.

We would still need to truncate some of the data if we tried to do that I think. So, I guess we could do that while giving a warning that data will be truncated.

**Review by alan-baker — APPROVED** [review]

**Pull request merged**